src/analysis/retail/shaman/enhancement/modules/resourcetracker/MaelstromWeaponSpenders.tsx

```diff
@@ -0,0 +1,152 @@
+import Analyzer, { Options, SELECTED_PLAYER } from 'parser/core/Analyzer';
+import MaelstromWeaponTracker from './MaelstromWeaponTracker';
+import ResourceBreakdown from 'parser/shared/modules/resources/resourcetracker/ResourceBreakdown';
+import Panel from 'parser/ui/Panel';
+import { MAELSTROM_WEAPON_ELIGIBLE_SPELLS } from '../../constants';
+import Events, { DamageEvent, HealEvent, RemoveBuffEvent } from 'parser/core/Events';
+import { SpellLink } from 'interface';
+import Abilities from 'parser/core/modules/Abilities';
+import { formatNumber, formatThousands } from 'common/format';
+import { TALENTS_SHAMAN } from 'common/TALENTS';
+import SPELLS, { maybeGetSpell } from 'common/SPELLS';
+import AbilityTracker from 'parser/shared/modules/AbilityTracker';
+import {
+  LIGHTNING_BOLT_PRIMORDIAL_WAVE_LINK,
+  PRIMORIDAL_WAVE_END_LINK,
+} from '../normalizers/EventLinkNormalizer';
+
+export default class extends Analyzer {
+  static dependencies = {
+    maelstromWeaponTracker: MaelstromWeaponTracker,
+    abilities: Abilities,
+    abilityTracker: AbilityTracker,
+  };
+
+  protected abilities!: Abilities;
+  protected abilityTracker!: AbilityTracker;
+  protected maelstromWeaponTracker!: MaelstromWeaponTracker;
+  protected spenderValues: Record<number, number> = [];
+  protected recordNextSpenderAmount: boolean = false;
+  protected maelstromSpendWithPrimordialWave: number = 0;
+
+  constructor(options: Options) {
+    super(options);
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER), this.onCast);
+
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(SPELLS.PRIMORDIAL_WAVE_BUFF),
+      this.onPrimoridalWaveEnd,
+    );
+
+    this.addEventListener(
+      Events.damage.by(SELECTED_PLAYER).spell(MAELSTROM_WEAPON_ELIGIBLE_SPELLS),
+      this.onSpender,
+    );
+    this.addEventListener(
+      Events.heal.by(SELECTED_PLAYER).spell(MAELSTROM_WEAPON_ELIGIBLE_SPELLS),
+      this.onSpender,
+    );
+  }
+
+  onCast() {
+    this.recordNextSpenderAmount = true;
+  }
+
+  onPrimoridalWaveEnd(event: RemoveBuffEvent) {
+    const primordialWaveLightningBolt = event._linkedEvents?.find(
+      (le) => le.relation === PRIMORIDAL_WAVE_END_LINK,
+    )?.event;
+    if (primordialWaveLightningBolt) {
+      const spent = this.maelstromWeaponTracker.current;
+      this.maelstromSpendWithPrimordialWave +=
+        spent === 0 ? 10 : this.maelstromWeaponTracker.current;
+      const damageEvents: DamageEvent[] = primordialWaveLightningBolt._linkedEvents
+        ?.filter((le) => le.relation === LIGHTNING_BOLT_PRIMORDIAL_WAVE_LINK)
+        .map((le) => le.event) as DamageEvent[];
```

Comment:
note that we have the `GetRelatedEvents` helper for cases like this (it is a little more awkward for cases with a single event like PW LB)

```suggestion
      const damageEvents: DamageEvent[] = GetRelatedEvents(primordialWaveLightningBolt, LIGHTNING_BOLT_PRIMORDIAL_WAVE_LINK) as DamageEvent[];
```

Reply:
Done